### PR TITLE
Extend extract function to support DateTime data types

### PR DESCRIPTION
**Pull request by mehant**
Currently extract() function only supports extracting time units from an Interval data type. This patch extends the extract() function to support DateTime.

In StandardConvertletTable() we reinterpret the extract() function to be a '/' expression based on the input data types. The problem is for Drill we cannot rewrite the extract() function this early because the data types are still 'ANY' and we cannot determine the multiplier to be used in the '/' expression while rewriting it.

Drill will extend the StandardConvertletTable() and override the convertExtract() method to simply treat it like a regular function and not rewrite it.

### CODE REVIEW

**Comment by julianhyde** [comment]
Can you add some tests to `SqlOperatorBaseTest` (maybe add a method below `testExtractFunc`). Then we can see that this function produces the right results.

File: core/src/main/java/net/hydromatic/optiq/tools/Frameworks.java [hunk 1]
```diff
@@ -113,12 +115,27 @@ public static Planner getPlanner(Lex lex,
...
+      RuleSet... ruleSets) {
+
     return new PlannerImpl(lex, parserFactory, schemaFactory, operatorTable,
         ImmutableList.copyOf(ruleSets),
-        traitDefs == null ? null : ImmutableList.copyOf(traitDefs));
+        traitDefs == null ? null : ImmutableList.copyOf(traitDefs),
+        convertletTable);
   }
 
-  /** Piece of code to be run in a context where a planner is available. The
```
Comment (julianhyde) [inline core/src/main/java/net/hydromatic/optiq/tools/Frameworks.java:121]: fix indent

File: core/src/main/java/net/hydromatic/optiq/tools/Frameworks.java [hunk 2]
```diff
@@ -113,12 +115,27 @@ public static Planner getPlanner(Lex lex,
       SqlOperatorTable operatorTable,
       List<RelTraitDef> traitDefs,
       RuleSet... ruleSets) {
+    return getPlanner(lex, parserFactory, schemaFactory, operatorTable,
+             traitDefs, new StandardConvertletTable(), ruleSets);
```
Comment (julianhyde) [inline core/src/main/java/net/hydromatic/optiq/tools/Frameworks.java]: please introduce a singleton, StandardConvertletTable.INSTANCE.

File: core/src/main/java/net/hydromatic/optiq/tools/Frameworks.java [hunk 3]
```diff
@@ -113,12 +115,27 @@ public static Planner getPlanner(Lex lex,
...
+  }
+
+  public static Planner getPlanner(Lex lex,
+      SqlParserImplFactory parserFactory,
```
Comment (julianhyde) [inline core/src/main/java/net/hydromatic/optiq/tools/Frameworks.java]: let's not have another getPlanner method. add a parameter to the existing method and compatibility be damned.

at some point we should introduce a builder, e.g.

``` java
Planner planner =
  Frameworks.plannerBuilder()
    .setLex(Lex.MYSQL)
    .setConvertletTable(StandardConvertletTable.INSTANCE)
    .build();
```

File: core/src/main/java/org/eigenbase/sql2rel/SqlToRelConverter.java [hunk 4]
```diff
@@ -144,6 +144,26 @@ public SqlToRelConverter(
       Prepare.CatalogReader catalogReader,
       RelOptPlanner planner,
       RexBuilder rexBuilder) {
+      this(viewExpander, validator, catalogReader, planner, rexBuilder,
+        new StandardConvertletTable());
+  }
+  /**
+   * Creates a converter.
```
Comment (julianhyde) [inline core/src/main/java/org/eigenbase/sql2rel/SqlToRelConverter.java:132]: don't create another constructor. add a parameter and fix up existing uses.

**Comment by mehant** [comment]
"Can you add some tests to SqlOperatorBaseTest (maybe add a method below testExtractFunc). Then we can see that this function produces the right results."

Did you want me to add negative tests, because StandardConvertletTable.convertExtract() will currently throw an exception if the second input is a DATETIME type. Or did you want me to modify StandardConvertletTable, because the current multipliers in convertExtract() wouldn't work with DATETIME types.

**Comment by julianhyde** [comment]
Can you write a test. If you can't easily get it to pass then you can add @Ignore.

In the long run, it will benefit you Drill developers, when you get round to calling SqlOperatorTest on your own operator implementations.

**Comment by mehant** [comment]
Thanks for the quick reply. I've added tests to SqlOperatorBaseTest. convertExtract() function in StandardConvertletTable will need to be enhanced to handle DATETIME types. I can create a separate issue to address this and work on it.

Made the other changes suggested by you.

**Pull request closed without merging**